common/finalizer_test.go

```diff
@@ -0,0 +1,134 @@
+
+	t.Run("register", func(t *testing.T) {
+		t.Run("succeeds", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			require.NoError(t, f.Register("1", nil))
+			require.NoError(t, f.Deregister("1"))
+			f.Run(1 * time.Second)
+		})
+
+		t.Run("fails when ID already registered", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			require.NoError(t, f.Register("1", nil))
+			require.ErrorIs(t, f.Register("1", nil), common.FinalizerDuplicateIdErr)
+		})
+
+		t.Run("fails after already run before", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			f.Run(1 * time.Second)
+			require.ErrorIs(t, f.Register("1", nil), common.FinalizerAlreadyDoneErr)
+		})
+	})
+
+	t.Run("deregister", func(t *testing.T) {
+		t.Run("succeeds", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			require.NoError(t, f.Register("1", nil))
+			require.NoError(t, f.Deregister("1"))
+			require.Zero(t, f.Run(1*time.Second))
+		})
+
+		t.Run("fails if callback does not exist", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			require.ErrorIs(t, f.Deregister("does-not-exist"), common.FinalizerUnknownIdErr)
+		})
+
+		t.Run("fails after already run before", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			f.Run(1 * time.Second)
+			require.ErrorIs(t, f.Deregister("1"), common.FinalizerAlreadyDoneErr)
+		})
+	})
+
+	t.Run("run", func(t *testing.T) {
+		t.Run("invokes all callbacks", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+
+			var completed atomic.Int32
+			for i := 0; i < 5; i += 1 {
+				require.NoError(t, f.Register(
+					fmt.Sprintf("%v", i),
+					func(ctx context.Context) error {
+						completed.Add(1)
+						//nolint:forbidigo
+						time.Sleep(50 * time.Millisecond)
+						return nil
+					}))
+			}
+
+			require.EqualValues(t, 5, f.Run(1*time.Second))
+			require.EqualValues(t, 5, completed.Load())
+		})
+
+		t.Run("returns once timeout has been reached", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			timeout := 50 * time.Millisecond
+
+			require.NoError(t, f.Register(
+				"before-timeout",
+				func(ctx context.Context) error {
+					return nil
+				}))
+
+			require.NoError(t, f.Register(
+				"after-timeout",
+				func(ctx context.Context) error {
+					//nolint:forbidigo
+					time.Sleep(2 * timeout)
+					return nil
+				}))
+
+			completed := f.Run(timeout)
+			require.EqualValues(t, 1, completed, "expected only one callback to complete")
+		})
+
+		t.Run("does not execute more than once", func(t *testing.T) {
+			f := common.NewFinalizer(log.NewNoopLogger())
+			require.NoError(t, f.Register(
+				"0",
+				func(ctx context.Context) error {
+					return nil
+				}))
+			require.EqualValues(t, 1, f.Run(1*time.Second))                            // 1st
+			require.Zero(t, f.Run(1*time.Second), "expected no callbacks to complete") // 2nd
+		})
+	})
+}
```

Comment:
NOTE: I decided to not re-test the adaptive pool behavior here again. This leaves a bit of a blind spot, but re-testing it seemed over the top.